### PR TITLE
DrivAerML: EMA=0.9995 + gc=0.5 + light WD (WD=1e-4 and WD=5e-4)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The EMA=0.9995+gc=0.5 champion (PR #3072, 3.833%) uses **no weight decay**. Light WD may improve generalization by preventing weight magnitudes from growing during the long training run, complementing EMA's exponential averaging effect. This tests whether WD and EMA are synergistic or antagonistic on DrivAerML.

Context from prior experiments:
- PR #3072 (eren): EMA=0.9995 + gc=0.5 + **no WD** → **3.833%** ← current champion
- PR #3046 (sukuna): WD=1e-3 + gc=1.0 + **no EMA** → 4.44% — WD alone doesn't help without EMA
- The question: does WD become beneficial when EMA is smoothing the weight trajectory?

Two WD levels to bracket the answer:
- **WD=1e-4** (very light) — gentle regularization, minimal interference with EMA
- **WD=5e-4** (moderate) — stronger regularization, more comparable to standard practice

Target: sub-3.82% (beat AB-UPT external baseline).

## Instructions

Run both variants sequentially. Use the same architecture and optimizer config as PR #3072 — the **only change** is adding `--weight-decay`.

**Variant 1: WD=1e-4 (very light)**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --weight-decay 1e-4 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --wandb-group dm-ema-wd-compound
```

**Variant 2: WD=5e-4 (moderate)**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --weight-decay 5e-4 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --wandb-group dm-ema-wd-compound
```

Note: `--wandb-group dm-ema-wd-compound` groups both runs together in W&B for easy comparison.

## Baseline

| Metric | Value | Source |
|--------|-------|--------|
| val_primary/surface_rel_l2_pct | **3.833%** | PR #3072, EMA=0.9995+gc=0.5, no WD, ep511 |
| test_primary/surface_rel_l2_pct | 4.685% | PR #3072, best-checkpoint eval |
| External target (AB-UPT) | <3.71% | ~500 epochs |

**Reproduce champion:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995
```

## What to report

For **each variant**, report:
- Best `val_primary/surface_rel_l2_pct` and at which epoch
- W&B run ID
- Compare vs 3.833% (no WD champion) and vs 4.44% (WD=1e-3+gc=1.0, no-EMA, PR #3046)
- Note: if either variant diverges early, report the divergence epoch and last stable val metric

Target: sub-**3.82%** on val to beat AB-UPT external baseline.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT emma:

## Results

### Variant 1: WD=1e-4 (very light)

| Metric | Value | Baseline (no WD) | Delta |
|--------|-------|-------------------|-------|
| val_primary/surface_rel_l2_pct | **14.76%** (ep35) | 3.833% (ep511) | +285% ❌ |
| test_primary/surface_rel_l2_pct | 15.43% | 4.685% | +229% ❌ |
| Divergence onset | ep36 | none through ep517 | — |
| grad_norm >100 | ep52 | never | — |

**W&B run:** cbkbmr7n (emma/dm-ema-wd1e4)

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 --weight-decay 1e-4 --enable-fourier --model-layers 4 \
  --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 \
  --agent emma --wandb-name "emma/dm-ema-wd1e4" --wandb-group dm-ema-wd-compound
```

**Training trajectory:** Smooth descent to 14.76% at ep35, then immediate divergence starting ep36 (+1.5pp), escalating through ep37 (+3.5pp), ep38 (+5.7pp). Grad norms first exceeded 100 at ep52, reached inf by ~ep60. Val metric flatlined at ~72% for remaining 450 epochs.

---

### Variant 2: WD=5e-4 (moderate)

| Metric | Value | Baseline (no WD) | Delta |
|--------|-------|-------------------|-------|
| val_primary/surface_rel_l2_pct | **4.920%** (ep247) | 3.833% (ep511) | +28.4% ❌ |
| test_primary/surface_rel_l2_pct | 5.853% | 4.685% | +24.9% ❌ |
| Divergence onset | ep248 | none through ep517 | — |
| grad_norm >100 | ep276 | never | — |

**W&B run:** t1pqa5qd (emma/dm-ema-wd5e4)

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 --weight-decay 5e-4 --enable-fourier --model-layers 4 \
  --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 \
  --agent emma --wandb-name "emma/dm-ema-wd5e4" --wandb-group dm-ema-wd-compound
```

**Training trajectory:** Smooth descent over 247 epochs (ep30=17.08%, ep60=10.33%, ep100=7.49%, ep200=5.48%). Best at ep247 (4.920%). Gradual rise post-best, then sharp gradient explosion at ep276 (grad_norm >100). Val metric flatlined at ~66% from ep300 onward.

---

### Comparison table

| Variant | Best Val | Best Epoch | Test | Divergence | vs Baseline |
|---------|----------|-----------|------|------------|-------------|
| **No WD (champion)** | **3.833%** | **511** | **4.685%** | **none** | — |
| WD=1e-4 | 14.76% | 35 | 15.43% | ep36 | +285% ❌ |
| WD=5e-4 | 4.920% | 247 | 5.853% | ep248 | +28.4% ❌ |
| WD=1e-3 no-EMA (#3046) | 4.44% | — | — | — | +15.8% ❌ |
| External (AB-UPT) | <3.71% | ~500 | — | — | target |

### What happened

**WD is antagonistic with EMA on DrivAerML.** Both WD variants underperform the no-WD champion, and both induce catastrophic divergence that the champion avoids entirely.

The mechanism appears to be: weight decay shrinks weights continuously, which disrupts the EMA shadow's ability to track a stable optimization trajectory. The EMA shadow becomes stale faster because the underlying weights are being actively decayed, creating a growing discrepancy between the live model and EMA. Eventually this mismatch triggers gradient instability that cascades into divergence.

Key evidence:
- **WD=1e-4 diverges much earlier (ep35) than WD=5e-4 (ep247)** — counterintuitively, the lighter regularization is worse. This is because WD=5e-4 provides enough regularization to constrain weight growth early, while WD=1e-4 is too weak to regularize but strong enough to destabilize the EMA trajectory.
- **The no-WD champion showed no divergence through 517 epochs**, confirming WD is the destabilizing factor, not the training length or LR schedule.
- **WD=1e-3 without EMA (#3046) achieved 4.44%** — WD alone isn't inherently bad, but the WD+EMA combination is.

### Bug fix note

Fixed a `primary_metric_key` shadowing bug in train.py (lines 1640-1643 in the original code) that caused an `UnboundLocalError` and crashed both initial run attempts. The local string variable `primary_metric_key` was shadowing the function of the same name defined at line 1423. Removed the redundant inline string assignment and replaced the downstream reference with `best_val_primary_metric_name` (already correctly set by the function call). This is a pre-existing bug in `main()`, not related to the WD experiment.

### Suggested follow-ups

1. **EMA warmup tuning for WD compatibility**: The EMA warmup formula may need adjustment when WD is present — slower warmup could help the EMA shadow establish before WD kicks in.
2. **WD scheduling**: Linearly ramping WD from 0 to target over the first 50-100 epochs might avoid the early destabilization observed in WD=1e-4.
3. **Decoupled WD**: AdamW already implements decoupled WD, but testing L2 regularization (coupled) might interact differently with EMA.
4. **The no-WD champion is robust**: This result strengthens the case that EMA=0.9995+gc=0.5 without WD is the right DrivAerML recipe. Further improvement should look elsewhere (LR schedule, architecture, longer training).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3153 because ADVISOR: WD dead on DM+EMA — both WD=1e-4 (14.76%, diverged ep36) and WD=5e-4 (4.920%, diverged ep248). WD destabilizes the EMA shadow model trajectory. No-WD confirmed for DM champion. Bug fix noted. Reassigning to noam-feature test on TandemFoil..
